### PR TITLE
Index select patch

### PR DESCRIPTION
**Pull request by jjsjann123**
Things in this PR:
1. fixes python frontend exposure of index_select.
2. add python frontend test for index_select.
3. cleans up boilerplate of scheduler for select/index_select/gather.
4. added fusion input check for select/index_select/gather ops in no-op scheduler. (Fixes part of #2352 on the inconsistent error message for slow tests. Note that the real fix also would requires the segment refactor in #2344 )

### CODE REVIEW

File: third_party/nvfuser/csrc/scheduler/registry.cpp [hunk 1]
```diff
@@ -34,24 +34,52 @@ std::deque<std::deque<TensorView*>> tvChains(
...
-        "First input of TorchGatherOp must be fusion input.");
+        (warning_message + " must be fusion input.").c_str());
     return true;
   }
-  if (torch_gather->input(0)->uses().size() > 1) {
+  if (op->input(0)->uses().size() > 1) {
     scheduler_debug_utils::canScheduleRejectReason(
         schedule_stragety,
-        "First input of TorchGatherOp can only be used by TorchGatherOp");
+        (warning_message + " can only be used by TorchGatherOp.").c_str());
```
Comment (zasdfgbnm) [inline third_party/nvfuser/csrc/scheduler/registry.cpp]: `TorchGatherOp` -> `op->getOpString()`

File: third_party/nvfuser/csrc/scheduler/registry.cpp [hunk 2]
```diff
@@ -34,24 +34,52 @@ std::deque<std::deque<TensorView*>> tvChains(
...
+    Op* op,
     ScheduleHeuristic schedule_stragety) {
-  if (!torch_gather->input(0)->isFusionInput()) {
+  std::string warning_message = "First input of ";
+  warning_message += op->getOpString();
+  if (!op->input(0)->isFusionInput()) {
     scheduler_debug_utils::canScheduleRejectReason(
         schedule_stragety,
-        "First input of TorchGatherOp must be fusion input.");
+        (warning_message + " must be fusion input.").c_str());
```
Comment (naoyam) [inline third_party/nvfuser/csrc/scheduler/registry.cpp]: Does it need to use `.c_str()`?

File: third_party/nvfuser/csrc/scheduler/registry.cpp [hunk 3]
```diff
@@ -34,24 +34,52 @@ std::deque<std::deque<TensorView*>> tvChains(
   return tv_chains;
 }
 
-bool rejectScheduleForTorchGather(
-    TorchGatherOp* torch_gather,
+// bool rejectScheduleForTorchGather(
```
Comment (naoyam) [inline third_party/nvfuser/csrc/scheduler/registry.cpp]: stale?

File: third_party/nvfuser/csrc/scheduler/registry.cpp [hunk 4]
```diff
@@ -34,24 +34,52 @@ std::deque<std::deque<TensorView*>> tvChains(
...
+template<typename Op>
+bool rejectScheduleFusionInputRequirement(
+    Op* op,
```
Comment (naoyam) [inline third_party/nvfuser/csrc/scheduler/registry.cpp]: Does `Op` need to be a template type?

File: third_party/nvfuser/csrc/scheduler/registry.cpp [hunk 5]
```diff
@@ -34,24 +34,52 @@ std::deque<std::deque<TensorView*>> tvChains(
...
+            idx_sel, schedule_strategy)) {
+      return true;
+    }
+  }
+  for (auto torch_gather : ir_utils::getTorchGatherOps(fusion)) {
+    if (rejectScheduleFusionInputRequirement(
+            torch_gather, schedule_strategy)) {
+      return true;
+    }
+  }
```
Comment (naoyam) [inline third_party/nvfuser/csrc/scheduler/registry.cpp]:
```suggestion
  for (auto expr: fusion->exprs()) {
    if (expr->isOneOf<SelectOp, IndexSelectOp, TorchGatherOp>() && 
      rejectScheduleFusionInputRequirement(
              expr, schedule_strategy)) {
      return true;
    }
  }
```

File: third_party/nvfuser/csrc/scheduler/registry.cpp (hunk 4 above)
Comment (jjsjann123) [reply third_party/nvfuser/csrc/scheduler/registry.cpp]: Ha this is a good catch!

I guess we are only using `getOpString` and `input` on these and I should have just used an Expr*.

File: third_party/nvfuser/csrc/scheduler/registry.cpp (hunk 2 above)
Comment (jjsjann123) [reply third_party/nvfuser/csrc/scheduler/registry.cpp]: Fancy, it's a variadic function, we don't even need to concat these strings!

**Review by naoyam — APPROVED** [review]
I'm stamping as the changes of registry.cpp look good, but the frontend changes should be reviewed by someone else.

**Comment by jjsjann123** [comment]
cc'ing @kevinstephano to go through python frontend patches.

I didn't really change anything, but the previous registration is just... broken. So I added the python test using index_select, which is now working~~ But I'll wait for Kevin's stamp before merging.

**Review by kevinstephano — APPROVED** [review]
The frontend changes look good!

**Pull request merged**